FirebaseInstallations/Source/Library/InstallationsIDController/FIRInstallationsIDController.m

```diff
@@ -33,47 +33,62 @@ @interface FIRInstallationsIDController ()
 
 @property(nonatomic, readonly) FIRInstallationsStore *installationsStore;
 
-// TODO: Use FIRInstallationsAPIService to register installation.
-//@property(nonatomic, readonly) FIRInstallationsAPIService *APIService;
+@property(nonatomic, readonly) FIRInstallationsAPIService *APIService;
+
+@property(atomic, strong, nullable) FBLPromise<FIRInstallationsItem *> *registrationPromise;
 @end
 
 @implementation FIRInstallationsIDController
 
 - (instancetype)initWithGoogleAppID:(NSString *)appID
                             appName:(NSString *)appName
-                             APIKey:(NSString *)APIKey {
+                             APIKey:(NSString *)APIKey
+                          projectID:(NSString *)projectID {
   FIRSecureStorage *secureStorage = [[FIRSecureStorage alloc] init];
   FIRInstallationsStore *installationsStore =
       [[FIRInstallationsStore alloc] initWithSecureStorage:secureStorage accessGroup:nil];
+  FIRInstallationsAPIService *apiService =
+      [[FIRInstallationsAPIService alloc] initWithAPIKey:APIKey projectID:projectID];
   return [self initWithGoogleAppID:appID
                            appName:appName
-                            APIKey:APIKey
-                installationsStore:installationsStore];
+                installationsStore:installationsStore
+                        APIService:apiService];
 }
 
 /// The initializer is supposed to be used by tests to inject `installationsStore`.
 - (instancetype)initWithGoogleAppID:(NSString *)appID
                             appName:(NSString *)appName
-                             APIKey:(NSString *)APIKey
-                 installationsStore:(FIRInstallationsStore *)installationsStore {
+                 installationsStore:(FIRInstallationsStore *)installationsStore
+                         APIService:(FIRInstallationsAPIService *)APIService {
   self = [super init];
   if (self) {
     _appID = appID;
     _appName = appName;
     _installationsStore = installationsStore;
+    _APIService = APIService;
   }
   return self;
 }
 
 - (FBLPromise<FIRInstallationsItem *> *)getInstallationItem {
```

Comment:
Should we also have a local cache here so it's not querying every time this is called? Considering it might access keychain.

Reply:
@chliangGoogle  Yeah, I was thinking of it. I would prefer to avoid the local cache if it is possible because managing it will make the logic more complicated and more vulnerable to the race conditions.

I'll make a note to measure  the performance of the current implementation to decide if dealing with the local cache is worth the efforts.

Reply:
I think we can only keep a few critical info that is read from storage to a central class where it's guarded by serial queue. Again as long as we keep it simple, it should not have race conditions ;P The issue with IID is there are twoo many classes scatter around and way too many keychain access especially during app start (many of them are repeat calling the same parameters, which causes long init time)

Reply:
@chliangGoogle Yeah, I see what you mean. I am thinking of adding a generic in-memory cache to `FIRSecureStorage`. Then the client code will stay the same and we will have to care of the implementation in a single place.
I'll do this in next PRs (probably after finishing the core functionality)